Print using callbacks

erick-xanadu opened the pull request:
**Context:** Use `debug.callback` to print.

**Description of the Change:** Creates callback to improve printing interface by providing the same semantics as JAX's debug print. Specializes the previous print implementation to memref.

**Benefits:** Better interface.

**Possible Drawbacks:** None.

**Related GitHub Issues:** None.

[sc-59388]

josh146 commented on doc/changelog.md:
````diff
@@ -26,6 +27,20 @@
+  ```
+
+  ```pycon
+  >>> print(func(1, 2))
+  2 1
````
```suggestion
  >>> func(1, 2)
  2 1
```

since 'console syntax' will always automatically print output :)

(technically, it will print the `repr()` rather than the `str()`, but same effect).

erick-xanadu replied:
From the testing, it looks like print uses string rather than repr : https://stackoverflow.com/a/36156152

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/09f95c5aa3ce6575dce1b3374b2e11a5e043c561

josh146 commented on frontend/catalyst/debug.py:
```diff
@@ -24,11 +26,34 @@
 from catalyst.utils.filesystem import WorkspaceManager
 
 
+def debug_print(fmt: str, *args, **kwargs) -> None:
+    """
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/5eeebf2c1829bf3c66c947b6e12464a779434e29/jax/_src/debugging.py#L269
+    version released under the Apache License, Version 2.0, with the following copyright notice:
+
+    Copyright 2022 The Jax Authors.
+    """
+    callback(partial(_format_print_callback, fmt))(*args, **kwargs)
```
My suggestion here is that, rather than a new function, we simply upstream this implementation into the existing `catalyst.debug.print`.

This might also solve a problem we currently have where `catalyst.debug.print` does not work in Jupyter notebooks.

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/b7ea323a46426dfe11390823172d2cdfbb46458e

erick-xanadu replied:
Also, I would like to suggest having two functions:

1. `print` similar to python's print
2. `printf` similar to what is implemented here.

Thoughts?

josh146 replied:
@erick-xanadu what is your motivation for the two functions?

erick-xanadu replied:
To keep the previous behaviour. They have different semantics. One is equivalent to the print statement. The other one is equivalent to: 

```python
def debug.print(fmt: str, *args: PyTree[Array], **kwargs: PyTree[Array]) -> None:
  print(fmt.format(*args, **kwargs))
```

josh146 replied:
yep! I guess the way I see it is from a user POV, it feels a bit like the same function 🤔 that is, `fmt` could simply be a string with no `{}`, and `*args` could be empty, and it acts like a print statement?

josh146 reviewed (changes requested):
thanks Erick, just a request to merge the two print functions!

josh146 commented on frontend/catalyst/debug.py:
```diff
@@ -16,63 +16,49 @@
-            base memory address of the data buffer, as well as the rank of the array, the size of
-            each dimension, and the strides between elements.
+def print(fmt: str, *args, **kwargs) -> None:
+    """
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/5eeebf2c1829bf3c66c947b6e12464a779434e29/jax/_src/debugging.py#L269
+    version released under the Apache License, Version 2.0, with the following copyright notice:
```
Let's please keep the original docstring :) But in addition, add documentation on how to use the format string syntax.

In addition @erick-xanadu, there is a question of should we keep the `memref` functionality. In which case, we can motivate keeping this implementation as `print_array_memref(x)`.

erick-xanadu replied:
I don't think there is a reason to keep the memref functionality unless team members are using it. @dime10 did you use it? Now you could implement printing out the address easily by doing the following, which in my opinion is not too bad to change when debugging and we can remove all of the other code.

```diff
diff --git a/frontend/catalyst/pennylane_extensions.py b/frontend/catalyst/pennylane_extensions.py
index 66248aac..fc757a6e 100644
--- a/frontend/catalyst/pennylane_extensions.py
+++ b/frontend/catalyst/pennylane_extensions.py
@@ -2474,6 +2474,7 @@ class CallbackClosure:
         jnpargs = []
         for void_ptr, ty in zip(flat_args, self.getLowLevelSignature):
             memref_ty = ctypes.cast(void_ptr, ty)
+            print(memref_ty)
             nparray = ranked_memref_to_numpy(memref_ty)
             jnparray = jnp.asarray(nparray)
             jnpargs.append(jnparray)
```

josh146 replied:
Got it, thanks @erick-xanadu. I have no strong opinion here, happy to keep it if it's useful/the team is using it, but also happy to not.

isaacdevlugt reviewed (changes requested):
- Functionality works the same as `jax.debug.print` ✅ 
- Can be used without additional arguments ✅ 
- old functionality has been moved to `catalyst.debug.print_memref` ✅ 
- `catalyst.debug`list in the docs has been updated, but `print` needs a description :)

<img width="818" alt="Screenshot 2024-04-25 at 9 04 26 PM" src="https://github.com/PennyLaneAI/catalyst/assets/34751083/ddd7cb33-b0be-4ee0-abcb-051737b45826">

erick-xanadu commented on doc/changelog.md:
```diff
@@ -151,6 +152,9 @@
 
 <h3>Breaking changes</h3>
 
+* The print function has changed. As a result, the memref keyword argument has been removed.
```
Changed here. Thanks! https://github.com/PennyLaneAI/catalyst/pull/621/commits/06f375e6667f47519fee81f196323a920c5c0b20

erick-xanadu commented on frontend/catalyst/callback/implementation.py:
```diff
@@ -75,47 +75,50 @@ def pure_callback(callback_fn, result_type=None):
-      2. The function has no side effect.
+    * Given the same arguments `*args`, the results will be the same each time the function is
+    called.
+    * The function has no side effect.
```
Added examples of side effects here: https://github.com/PennyLaneAI/catalyst/pull/621/commits/06f375e6667f47519fee81f196323a920c5c0b20#diff-fc94e32bec1cd52135a8417c323c1eb7c072bef627072aa02680a3144017a4e1

erick-xanadu commented:
> catalyst.debuglist in the docs has been updated, but print needs a description :)

@isaacdevlugt do you have a suggestion for that short line? I added more documentation but the user needs to click on the link to see it. I have modified to the following:

```
    """debug.callback : useful for printing and logging
```

erick-xanadu commented on doc/changelog.md:
```diff
@@ -9,30 +9,50 @@
-  @pure_callback
+  @debug.callback
   def foo(val):
     return val
```
```suggestion
    print(val)
```

dime10 commented on doc/changelog.md:
````diff
@@ -9,30 +9,50 @@
 
   ```py
-  @pure_callback
+  @debug.callback
   def foo(val):
     return val
````
This doesnt showcase the "debug" aspect very well, maybe an f-string print would be nice?

erick-xanadu replied:
Ah yes I did `git push -f` after the update and didn't update what I had done in the online code editor.

dime10 commented on doc/changelog.md:
```diff
@@ -9,30 +9,50 @@
   while `debug.callback` is expected to produce a side effect and have no return values.
 
+  Syntactic sugar on top of `debug.callback` includes `debug.print` which allows the user
+  to use f-stringlike to format the string before printing.
```
I would compare it `str.format()` rather than fstrings because its behaviour is closer to that right? 
`print("{c} {b} {a}", a=a, b=b, c=c)` ~ `print("{c} {b} {a}".format(a=a, b=b, c=c))`

dime10 replied:
Might even make sense to link to the doc page: https://docs.python.org/3/library/string.html#formatstrings

erick-xanadu replied:
```suggestion
  to use formatstrings to format the string before printing. See https://docs.python.org/3/library/string.html#formatstrings
```

dime10 commented on frontend/catalyst/cuda/primitives/__init__.py:
```diff
@@ -22,7 +22,7 @@
 # We disable protected access in particular to avoid warnings with cudaq._pycuda.
 # And we disable unused-argument to avoid unused arguments in abstract_eval, particularly kwargs.
-# pylint: disable=protected-access,unused-argument,abstract-method,line-too-long
+# pylint: disable=protected-access,unused-argument,line-too-long
```
Are you not getting the pylint warnings for this when you run pylint locally?

erick-xanadu replied:
Running locally? :sunglasses:

erick-xanadu replied:
Haha, all I know is that pylint online complained that this was useless.

```suggestion
# pylint: disable=protected-access,unused-argument,abstract-method,line-too-long
```

dime10 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,40 +15,77 @@
     Args:
-        x (jax.Array, Any): A single jax array whose numeric values are printed at runtime, or any
-            object whose string representation will be treated as a constant and printed at runtime.
-        memref (Optional[bool]): When set to ``True``, additional information about how the array is
```
We should preserve the description of this, even if the flag is removed since that is now the behaviour of the function.

erick-xanadu replied:
I didn't remove it, it is on `print_memref` below.

erick-xanadu replied:
Ah, I am not entirely sure what you mean here. You just want everything after `memref (Optional[bool])` to be included in the docstring?

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/186ab31134a57e8d86cd4f0296fea4286c3d65b6

dime10 replied:
Yep that's what I meant, although it shouldn't be listed under the `x` arg, just in the general docstring :)

erick-xanadu replied:
https://github.com/PennyLaneAI/catalyst/pull/621/commits/fb8132728b07accacbbb1dde1e5bd5b832d03afe

dime10 commented on frontend/catalyst/jax_primitives.py:
```diff
@@ -76,7 +76,7 @@
 from catalyst.utils.types import convert_shaped_arrays_to_tensors
 
-# pylint: disable=unused-argument,abstract-method,too-many-lines
+# pylint: disable=unused-argument,too-many-lines
```
same here, I'm getting tons of them which is why i added it

erick-xanadu replied:
```suggestion
# pylint: disable=unused-argument,abstract-method,too-many-lines
```

erick-xanadu commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,40 +15,78 @@
 """This module enables runtime printing of program values."""
 
 import builtins
+import sys
```
```suggestion
```

erick-xanadu commented on doc/changelog.md:
````diff
@@ -9,30 +9,50 @@
   ```pycon
   >>> print(circuit(123))
-  123
+  124
````
```suggestion
myval: 124
```

josh146 commented on doc/changelog.md:
````diff
@@ -9,30 +9,50 @@
+  Syntactic sugar on top of `debug.callback` includes `debug.print` which allows the user
+  to use f-stringlike to format the string before printing.
+
+  ```py
+  @qjit
+  def cir(a, b, c):
+      debug.print("{c} {b} {a}", a=a, b=b, c=c)
+  ```
+
+  ```pycon
+  >>> cir(1, 2, 3)
+  3 2 1
+  ```
````
This will be a massive QOL improvement 😎

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,40 +15,77 @@
 # pylint: disable=redefined-builtin
-def print(x, memref=False):
+def print(fmt, *args, **kwargs):
+    """A print function.
```
```suggestion
    """A :func:`~.qjit` compatible print function for printing values at runtime.
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,90 @@
+def print(fmt, *args, **kwargs):
+    """A print function.
+
+    This print function works with python callbacks, so all arguments will be printed in the same
+    way as the :func:`builtins.print` function prints them.
+
+    One difference is that if the first argument is a string, it may be used similarly to an
+    fstring. Like so:
+
+    .. code-block:: python
+
+        @qjit
+        def cir(a, b, c):
+            debug.print("c={c} b={b} a={a}", a=a, b=b, c=c)
+
+        cir(1, 2, 3)
+    >>> c=3 b=2 a=1
```
```suggestion
    This print function allows printing of values at runtime, unlike usage of standard Python ``print`` which will print values at program capture/compile time.
    
    ``debug.print`` is a minimal wrapper around :func:`~.debug.callback` which calls Python's ``builtins.print`` function, and thus will use the same formatting styles as Python's built-in print function.
    
    Args:
        fmt (str): The string to be printed. Note that this may also be a
            format string used to format input arguments (for example 
            ``cost={x}``), similar to those permitted by ``str.format``. See 
            the Python docs on
            `string formatting <https://docs.python.org/3/library/stdtypes.html#str.format>`__
            and `format string syntax <https://docs.python.org/3/library/string.html#formatstrings>`__.
        **args: Arguments to be passed to the format string.
        **kwargs: Keyword arguments to be passed to the format string.
        
    .. seealso:: :func:`~.print_memref`

    **Example**

    .. code-block:: python

        @qjit
        def f(a, b, c):
            debug.print("c={c} b={b} a={a}", a=a, b=b, c=c)

    >>> f(1, 2, 3)
    c=3 b=2 a=1
    
    .. note::
    
        Using Python f-strings as the `fmt` string will not work as expected since they will be treated as Python objects.
        This means that array values embedded in them will have their compile-time representation
        printed, instead of actual data.
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,90 @@
+    >>> c=3 b=2 a=1
+    """
+
+    if isinstance(fmt, str):
+        debug_callback(partial(_format_print_callback, fmt))(*args, **kwargs)
+        return
+
+    debug_callback(_print_callback)(fmt, *args, **kwargs)
```
@erick-xanadu this is my fault for missing this, but I see that the old print function had this behaviour:

> Outside a qjit() compiled function the operation falls back to the Python print statement.

I didn't include this in the requirements as I had forgotten this. I'm curious if the new function already works like this?

josh146 replied:
If not, no worries; I can add this as a new feature on SC

erick-xanadu replied:
It works!

```python
from catalyst.debug import print
print("hello world")
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,90 @@
+
+# pylint: disable=redefined-builtin
+def print_memref(x):
     """A :func:`qjit` compatible print function for printing values at runtime.
```
```suggestion
    """A :func:`qjit` compatible print function for printing numeric values at runtime with memref information.
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,90 @@
+            For more information about memref descirptors visit:
+            https://mlir.llvm.org/docs/Dialects/MemRef/
+            This includes the base memory address of the data buffer, as well as the rank of
+            the array, the size of each dimension, and the strides between elements.
```
```suggestion
            the array, the size of each dimension, and the strides between elements.
            
.. seealso:: :func:`~.debug.print`
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,90 @@
     [0.43]
-    exit
 
     Outside a :func:`qjit` compiled function the operation falls back to the Python print statement.
```
```suggestion
    Outside a :func:`qjit` compiled function the operation falls back to the Python print statement.
    
    .. note::
    
        Python f-strings will not work as expected since they will be treated as Python objects.
        This means that array values embedded in them will have their compile-time representation
        printed, instead of actual data.
        
        For printing format strings at *runtime*, please use :func:`~.debug.print`.
```

erick-xanadu replied:
I feel like this one is no longer relevant since this function is not expected to be compatible with strings nor f-strings?

josh146 reviewed (changes requested):
Nice, thanks @erick-xanadu! Mostly documentation changes

dime10 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,114 @@
+
+    Enables printing of numeric values at runtime and the value's metadata.
+
+    Tensors, in the Catalyst runtime are represented as memref descriptor structs.
+    For more information about memref descirptors visit:
```
```suggestion
    Tensors in the Catalyst runtime are represented as memref descriptor structs.
    For more information about memref descriptors visit:
```

dime10 reviewed (approved):
Looks good to me 👍

isaacdevlugt reviewed (approved):
Thanks @erick-xanadu!

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,115 @@
+
+    .. note::
+
+        Using Python f-strings as the `fmt` string will not work as expected since they will be
```
```suggestion
        Using Python f-strings as the ``fmt`` string will not work as expected since they will be
```

josh146 commented on frontend/catalyst/debug/printing.py:
```diff
@@ -15,54 +15,115 @@
+
+    Enables printing of numeric values at runtime and the value's metadata.
+
+    Tensors in the Catalyst runtime are represented as memref descriptor structs.
+    For more information about memref descriptors visit:
+    https://mlir.llvm.org/docs/Dialects/MemRef/
```
```suggestion
    Tensors in the Catalyst runtime are represented as memref descriptor structs.
    For more information about memref descriptors see the `MLIR documentation <https://mlir.llvm.org/docs/Dialects/MemRef/>`__.
```

josh146 reviewed (approved):
amazing, thanks @erick-xanadu
